lib/Dialect/FIRRTL/Transforms/LowerGroups.cpp

```diff
@@ -0,0 +1,391 @@
+        builder.create<StrictConnectOp>(newModule.getPortLocationAttr(portNum),
+                                        instanceOp.getResult(portNum),
+                                        connectValues[portNum]);
+      else
+        builder.create<StrictConnectOp>(
+            getPortLoc(connectValues[portNum]), connectValues[portNum],
+            builder.create<RefResolveOp>(newModule.getPortLocationAttr(portNum),
+                                         instanceOp.getResult(portNum)));
+    }
+    group.erase();
+
+    return WalkResult::advance();
+  });
+}
+
+/// Process a circuit to remove all groups in each module and top-level group
+/// declarations.
+void LowerGroupsPass::runOnOperation() {
+  LLVM_DEBUG(
+      llvm::dbgs() << "==----- Running LowerGroups "
+                      "-------------------------------------------------===\n");
+  CircuitOp circuitOp = getOperation();
+
+  // Initialize members which cannot be initialized automatically.
+  llvm::sys::SmartMutex<true> mutex;
+  circuitMutex = &mutex;
+  CircuitNamespace ns(circuitOp);
+  circuitNamespace = &ns;
+
+  // Lower the groups of each module.
+  SmallVector<FModuleOp> modules(circuitOp.getBodyBlock()->getOps<FModuleOp>());
+  llvm::parallelForEach(modules,
+                        [&](FModuleOp moduleOp) { runOnModule(moduleOp); });
+
+  // Generate the header and footer of each bindings file.  The body will be
+  // populated later when binds are exported to Verilog.  This produces text
+  // like:
+  //
+  //     `include "groups_A.sv"
+  //     `include "groups_A_B.sv"
+  //     `ifndef groups_A_B_C
+  //     `define groups_A_B_C
+  //     <body>
+  //     `endif // groups_A_B_C
+  //
+  // TODO: This would be better handled without the use of verbatim ops.
+  OpBuilder builder(circuitOp);
+  SmallVector<std::pair<GroupDeclOp, StringAttr>> groupDecls;
+  StringRef circuitName = circuitOp.getName();
+  circuitOp.walk<mlir::WalkOrder::PreOrder>([&](GroupDeclOp groupDeclOp) {
+    auto parentOp = groupDeclOp->getParentOfType<GroupDeclOp>();
+    while (parentOp && parentOp != groupDecls.back().first)
+      groupDecls.pop_back();
+    builder.setInsertionPointToStart(circuitOp.getBodyBlock());
+
+    // Save the "groups_CIRCUIT_GROUP" string as this is reused a bunch.
+    SmallString<32> prefix("groups_");
+    prefix.append(circuitName);
+    prefix.append("_");
+    for (auto [group, _] : groupDecls) {
+      prefix.append(group.getSymName());
+      prefix.append("_");
+    }
+    prefix.append(groupDeclOp.getSymName());
+
+    auto outputFileAttr = hw::OutputFileAttr::getFromFilename(
+        builder.getContext(), prefix + ".sv",
+        /*excludeFromFileList=*/true);
+
+    SmallString<128> includes;
+    for (auto [_, strAttr] : groupDecls) {
+      includes.append(strAttr);
+      includes.append("\n");
+    }
+
+    // Write header to a verbatim.
+    builder
+        .create<sv::VerbatimOp>(groupDeclOp.getLoc(), includes + "`ifndef " +
```

Comment:
:(

Reply:
I will plan to revisit this in later commits.